bgpd/bgp_mplsvpn.c

```diff
@@ -1409,6 +1457,172 @@ void vpn_policy_routemap_event(const char *rmap_name)
+
+	export_name = to_bgp->name ? to_bgp->name : BGP_DEFAULT_NAME;
+	tmp_name = from_bgp->name ? from_bgp->name : BGP_DEFAULT_NAME;
+	idir = BGP_VPN_POLICY_DIR_FROMVPN;
+	edir = BGP_VPN_POLICY_DIR_TOVPN;
+
+	/* Were we importing from "import_vrf"? */
+	for (ALL_LIST_ELEMENTS_RO(to_bgp->vpn_policy[afi].import_vrf, node,
+				  vname)) {
+		if (strcmp(vname, tmp_name) == 0)
+			break;
+	}
+
+	/*
+	 * We do not check in the cli if the passed in bgp
+	 * instance is actually imported into us before
+	 * we call this function.  As such if we do not
+	 * find this in the import_vrf list than
+	 * we just need to return safely.
+	 */
+	if (!vname)
+		return;
+
+	/* Remove "import_vrf" from our import list. */
+	listnode_delete(to_bgp->vpn_policy[afi].import_vrf, vname);
+	XFREE(MTYPE_TMP, vname);
+
+	/* Remove routes imported from "import_vrf". */
+	/* TODO: In the current logic, we have to first remove all
+	 * imported routes and then (if needed) import back routes
+	 */
+	vpn_leak_prechange(idir, afi, bgp_get_default(), to_bgp);
+
+	if (to_bgp->vpn_policy[afi].import_vrf->count == 0) {
+		UNSET_FLAG(to_bgp->af_flags[afi][safi],
+			   BGP_CONFIG_VRF_TO_VRF_IMPORT);
+		ecommunity_free(&to_bgp->vpn_policy[afi].rtlist[idir]);
+	} else {
+		ecom = from_bgp->vpn_policy[afi].rtlist[edir];
+		ecommunity_del_val(to_bgp->vpn_policy[afi].rtlist[idir],
+				   (struct ecommunity_val *)ecom->val);
+		vpn_leak_postchange(idir, afi, bgp_get_default(), to_bgp);
+	}
+
+	/*
+	 * What?
+	 * So SA is assuming that since the ALL_LIST_ELEMENTS_RO
+	 * below is checking for NULL that export_vrf can be
+	 * NULL, consequently it is complaining( like a cabbage )
+	 * that we could dereference and crash in the listcount(..)
+	 * check below.
+	 * So make it happy, under protest, with liberty and justice
+	 * for all.
+	 */
+	assert(from_bgp->vpn_policy[afi].export_vrf);
+
+	/* Remove us from "import_vrf's" export list. If no other VRF
+	 * is importing from "import_vrf", cleanup appropriately.
+	 */
+	for (ALL_LIST_ELEMENTS_RO(from_bgp->vpn_policy[afi].export_vrf,
+				  node, vname)) {
+		if (strcmp(vname, export_name) == 0)
+			break;
+	}
+
+	/*
+	 * If we have gotten to this point then the vname must
+	 * exist.  If not, we are in a world of trouble and
+	 * have slag sitting around.
+	 *
+	 * import_vrf and export_vrf must match in having
+	 * the in/out names as appropriate.
+	 */
+	assert(vname);
+
+	listnode_delete(from_bgp->vpn_policy[afi].export_vrf, vname);
+	XFREE(MTYPE_TMP, vname);
+
+	if (!listcount(from_bgp->vpn_policy[afi].export_vrf)) {
+		vpn_leak_prechange(edir, afi, bgp_get_default(), from_bgp);
+		ecommunity_free(&from_bgp->vpn_policy[afi].rtlist[edir]);
+		UNSET_FLAG(from_bgp->af_flags[afi][safi],
+			   BGP_CONFIG_VRF_TO_VRF_EXPORT);
+		memset(&from_bgp->vpn_policy[afi].tovpn_rd, 0,
+		       sizeof(struct prefix_rd));
+		UNSET_FLAG(from_bgp->vpn_policy[afi].flags,
+			   BGP_VPN_POLICY_TOVPN_RD_SET);
+		from_bgp->vpn_policy[afi].tovpn_label = 0;
```

Comment:
shouldn't this be MPLS_LABEL_NONE not 0?

Reply:
done